File: go/who_took.go
package main

func whoTook(
	cards []DownCard,
	trumpColor CardColor,
	firstCardIndex int,
) int {
	jokerValues := rotateArray([]int{0, 1, 2, 3}, firstCardIndex)

	firstCard := cards[firstCardIndex]
	firstCardColor := getNatarebiColor(firstCard)

	jokerWantsNonTrump := false
	var jokerDoesNotWantColor *CardColor

	if action := firstCard.JokerSubAction; action != nil {
		if action.Want && *action.Color != trumpColor {
			jokerWantsNonTrump = true
		}

		if !action.Want {
			jokerDoesNotWantColor = action.Color
		}
	}

	cardValues := []int{}

	for i, card := range cards {
		color := card.Color
		level := card.Level
		jokerAction := card.JokerSubAction
		isFirstCard := i == firstCardIndex

		if jokerAction != nil {
			if jokerAction.Want {
				priority := 100

				if isFirstCard && jokerWantsNonTrump {
					priority = 20
				}

				cardValues = append(cardValues, priority+jokerValues[i])
				continue
			}

			if isFirstCard {
				cardValues = append(cardValues, 1)
				continue
			}

			cardValues = append(cardValues, 0)
			continue
		}

		// no joker

		isTrumpColor := color == trumpColor
		isFirstCardColor := color == firstCardColor

		if !isFirstCard && !isFirstCardColor && !isTrumpColor {
			cardValues = append(cardValues, 0)
			continue
		}

		isJokerDoesNotWantColor := false

		if jokerDoesNotWantColor != nil && color == *jokerDoesNotWantColor {
			isJokerDoesNotWantColor = true
		}

		priority := level + 1

		if isTrumpColor {
			priority += 50
		}

		if isJokerDoesNotWantColor {
			priority += 20
		}

		cardValues = append(cardValues, int(priority))
	}

	return indexOf(cardValues, max(cardValues))
}

File: ts/src/who-took.ts
import { CardColor, DownCard } from './types';
import { getNatarebiColor, rotateArray } from './utils';

export function whoTook(
  cards: DownCard[],
  trumpColor: CardColor,
  firstCardIndex: number,
) {
  const jokerValues = rotateArray([0, 1, 2, 3], firstCardIndex);

  const firstCard = cards[firstCardIndex];
  const firstCardColor = getNatarebiColor(firstCard);

  // როცა ჯოკერმა აცხადა რომ სურს კოზირისგან განსხვავებული ფერი
  const jokerWantsNonTrump =
    firstCard[2]?.want && firstCard[2].color !== trumpColor;

  const jokerDontWantColor =
    firstCard[2] && !firstCard[2].want ? firstCard[2].color : null;

  const cardValues = cards.map(function (
    [color, level, jokerAction],
    i,
  ) {
    return jokerAction
      ? // თუ ჯოეკრია
        jokerAction.want
        ? // თუ სურს ჯოკერს
          (jokerWantsNonTrump && i === firstCardIndex ? 20 : 100) +
          // ბოლო ჩამოსულ ჯოკერს მეტი უპირატესობა აქვს
          jokerValues[i]
        : // თუ არ სურს ჯოკერს
        i === firstCardIndex
        ? 1
        : 0
      : // თუ არც ნატარები ფერია და არც კოზირი, მაშინ ვანულებთ
      i !== firstCardIndex &&
        color !== firstCardColor &&
        color !== trumpColor
      ? 0
      : level +
        // რადგან level-ი 0-დან იწყება
        1 +
        // კოზირს დამატებით 20 ქულა უპირატესობის მოსაპოვებლად
        // რადგან ფერში 9 კარტია მაქსიმუმ +20 ქულა საკმარისია
        (color === trumpColor ? 50 : 0) +
        // როდესაც ჯოკერმა აცხადა წაიღოსო და
        // კარტის ფერი დაემთხვა ნაცხადებ ფერს
        (color === jokerDontWantColor ? 20 : 0);
  });

  const max = Math.max(...cardValues);

  return cardValues.indexOf(max);
}
